### PR TITLE
add fp16 training code

### PR DESCRIPTION
**Pull request by danleifeng**
add fp16 training code

### CODE REVIEW

File: plsc/models/dist_algo.py [hunk 1]
```diff
@@ -33,16 +38,77 @@ class DistributedClassificationOptimizer(Optimizer):
...
+                 use_fp16=False,
+                 loss_type='dist_arcface',
+                 amp_lists=None,
+                 init_loss_scaling=1.0,
+                 incr_every_n_steps=1000,
+                 decr_every_n_nan_or_inf=2,
+                 incr_ratio=2.0,
+                 decr_ratio=0.5,
+                 use_dynamic_loss_scaling=True):
+        super(DistributedClassificationOptimizer, self).__init__(
```
[the remark this answers is not in the document]
Comment (danleifeng) [reply plsc/models/dist_algo.py]: done. thanks.

File: plsc/models/dist_algo.py [hunk 2]
```diff
@@ -59,54 +125,222 @@ def minimize(self,
...
+                    with switch.case(is_overall_finite):
+                        pass
+                    with switch.default():
+                        for _, g in scaled_params_grads:
+                            layers.assign(layers.zeros_like(g), g)
+
+            optimize_ops = self._optimizer.apply_gradients(scaled_params_grads)
+            ret = optimize_ops, scaled_params_grads
+
         return ret
```
[the remark this answers is not in the document]
Comment (danleifeng) [reply plsc/models/dist_algo.py]: done. thanks.

File: plsc/models/dist_algo.py [hunk 3]
```diff
@@ -33,16 +38,74 @@ class DistributedClassificationOptimizer(Optimizer):
...
+                 batch_size,
+                 use_fp16=False,
+                 loss_type='dist_arcface',
+                 amp_lists=None,
+                 init_loss_scaling=1.0,
+                 incr_every_n_steps=1000,
+                 decr_every_n_nan_or_inf=2,
+                 incr_ratio=2.0,
+                 decr_ratio=0.5,
+                 use_dynamic_loss_scaling=True):
```
Comment (gavin1332) [inline plsc/models/dist_algo.py]: 参数集合是开放集合吗？如果是开放集合是否应该传一个dict进来更好？

File: plsc/models/dist_algo.py [hunk 4]
```diff
@@ -33,16 +38,74 @@ class DistributedClassificationOptimizer(Optimizer):
...
         self._use_fp16 = use_fp16
+        self._amp_lists = amp_lists
+        if amp_lists is None:
+            self._amp_lists = AutoMixedPrecisionLists()
+
+        self._param_grads = None
+        self._scaled_loss = None
+        self._loss_type = loss_type
+        self._init_loss_scaling = init_loss_scaling
+        self._loss_scaling = layers.create_global_var(
```
Comment (gavin1332) [inline plsc/models/dist_algo.py]: 无论是否FP16训练，这些参数都会被初始化吗？

File: plsc/models/dist_algo.py [hunk 5]
```diff
@@ -59,55 +122,222 @@ def minimize(self,
...
+                index + 1,
+                type='scale',
+                inputs={'X': shard_logit_grad},
+                outputs={'Out': shard_logit_grad},
+                attrs={
+                    'scale': 1.0 / self._batch_size,
+                    op_role_key: loss_backward_role
+                })
+            return ret
+        else:
```
Comment (gavin1332) [inline plsc/models/dist_algo.py]: else block代码量过大，逻辑也比较多，适合抽出一个单独功能的函数，在此处调用。代码可读性更好。

File: plsc/models/dist_algo.py [hunk 6]
```diff
@@ -59,55 +122,222 @@ def minimize(self,
...
+                    index,
+                    type="cast",
+                    inputs={"X": shard_one_hot},
+                    outputs={"Out": shard_one_hot_fp32},
+                    attrs={
+                        "in_dtype": fluid.core.VarDesc.VarType.FP16,
+                        "out_dtype": fluid.core.VarDesc.VarType.FP32,
+                        op_role_key: backward_role
+                    })
+                name = 'tmp_3@GRAD'
```
Comment (gavin1332) [inline plsc/models/dist_algo.py]: 这个名字是怎么来的 ？

File: plsc/models/dist_algo.py [hunk 7]
```diff
@@ -59,55 +122,222 @@ def minimize(self,
...
+                assert block.ops[index + 3].type == 'elementwise_add_grad'
+ 
+                block._remove_op(index + 1)
+                block._remove_op(index)
+                block._remove_op(index - 1)
+
+                # insert the calculated gradient 
+                shard_one_hot = fluid.layers.create_tensor(
+                    fp32, name='shard_one_hot')
+                shard_one_hot_fp32 = fluid.layers.create_tensor(fp32,
```
Comment (gavin1332) [inline plsc/models/dist_algo.py]: 如何确保名称不冲突呢？有使用过unique_name模块吗

File: plsc/models/dist_algo.py [hunk 8]
```diff
@@ -59,55 +122,222 @@ def minimize(self,
...
+                block._remove_op(index - 1)
+
+                # insert the calculated gradient 
+                shard_one_hot = fluid.layers.create_tensor(
+                    fp32, name='shard_one_hot')
+                shard_one_hot_fp32 = fluid.layers.create_tensor(fp32, 
+                    name=(shard_one_hot.name+".cast_fp32"))
+                shard_logit_grad_fp32 = block.var(
+                    shard_logit.name+".cast_fp32@GRAD")
+                block._insert_op(
```
Comment (gavin1332) [inline plsc/models/dist_algo.py]: dist_softmax和dist_arcface分支大量重复代码，需要提炼一下

File: plsc/models/dist_algo.py (hunk 5 above)
Comment (danleifeng) [reply plsc/models/dist_algo.py]: 已修改，thanks！

File: plsc/models/dist_algo.py (hunk 7 above)
Comment (danleifeng) [reply plsc/models/dist_algo.py]: shard_one_hot name在fp32 和fp16下均未使用过，打印改变前后的program有确认过。shard_one_hot.cast_fp32只是根据你cast的命令规则create的一个tensor。

Comment (danleifeng) [reply plsc/models/dist_algo.py]: 已修改为unique_name,thanks!

File: plsc/models/dist_algo.py (hunk 3 above)
Comment (gavin1332) [reply plsc/models/dist_algo.py]: FP16的参数最好以一个特有的数据结构传进来，把FP16参数跟其他通用参数比如optimizer，batch size混在一起，用户难以理解。

File: plsc/models/dist_algo.py (hunk 8 above)
Comment (gavin1332) [reply plsc/models/dist_algo.py]: “index数，remove的个数，输入输出”这些不能作为参数传递吗？

File: plsc/entry.py [hunk 9]
```diff
@@ -293,8 +308,12 @@ def _get_optimizer(self):
     
         if self.loss_type in ["dist_softmax", "dist_arcface"]:
             self.optimizer = DistributedClassificationOptimizer(
-                self.optimizer, global_batch_size)
-
+                self.optimizer, global_batch_size, use_fp16=self.use_fp16,
+                loss_type=self.loss_type,
+                fp16_user_dict=self.fp16_user_dict)
+        elif self.use_fp16:
```
Comment (gavin1332) [inline plsc/entry.py]: how to use both dist_xxx and use_fp16?

File: plsc/models/dist_algo.py [hunk 10]
```diff
@@ -23,62 +23,260 @@
...
+        fp16_params_dict['incr_ratio'] = 2.0
+        fp16_params_dict['decr_ratio'] = 0.5
+        fp16_params_dict['use_dynamic_loss_scaling'] = True
+        if fp16_user_dict is not None:
+            # update fp16_params_dict
+            for key in fp16_user_dict:
+                if fp16_params_dict.has_key(key):
+                    fp16_params_dict[key] = fp16_user_dict[key]
+                else:
+                    print("Warning: can't find name %s in our fp16_params_dict. "
```
Comment (gavin1332) [inline plsc/models/dist_algo.py]: is it better to use logging?

File: plsc/entry.py (hunk 9 above)
Comment (danleifeng) [reply plsc/entry.py]: if use dist_xxx, DistributedClassificationOptimizer will be used no matter when use_fp16 equals True or not.
if use other loss type, decorate function will be called if use_fp16 equals True.

File: plsc/models/dist_algo.py (hunk 10 above)
Comment (danleifeng) [reply plsc/models/dist_algo.py]: done.thanks!
